### ISSUE TITLE
make publishing version format configurable from semantic-release cli

### ISSUE THREAD

**Issue by tarazansafak** [semantic-release/semantic-release#681]
## New feature motivation

In some cases like releasing from multiple branches, it would be nice to have possibility to have control over the published artifact's version's prefix for releasing from feature branch purpose.

## New feature description

--publish-version-format xyz-\${version}
where xyz-${version} is the version which will be written in to package.json

## New feature use case:

I would like to publish from multiple branches:

Run something similar on each branch via multibranch pipeline:

semantic-release --tag-format xyz-\${version} --publish-version-format xyz-\${version}
npm dist-tag add <pkg>@xyz-\${version} $BRANCH_NAME

By this way, we can have a component X tagged with feature branch version on Git and in NPM registry. The latest dist-tag will still show to latest of the released version not feature branch version. Thus we can just add another step to dist-tag the component X with branch name.

Basically, I would like to get some control from cli to interfere "prepare" step. So that I can append some things to the version which is going to be published to NPM registry.

**Comment by pvdlg** [comment]
I'm not really sure what the feature you are proposing would do exactly.
Can you add more clarifications?

Also, would it be covered by #563?

**Comment by tarazansafak** [comment]
I have updated my description and I was wondering should I try to write a prepare plugin and do my trick there? So that I can update the package.json with desired formatted version and publish

**Comment by pvdlg** [comment]
Can you clarify what you mean by "the published artifact's version's prefix"?
npm and other package management registry doesn't have any version prefixes. Your version has to be a valid semver like `1.0.0` not `xyz-1.0.0`.

npm support dist-tag, which allow you to make a given version available only to certain users. Releasing on different dist-tag based on the release branch is what is currently developed as part of #563.

**Comment by tarazansafak** [comment]
I got it working just today by using semantic-release/exec plugin and today I am publish in this format:

packageName-semver-customPostfix ie:
fdp-tool-cli-1.1.0-feature14

You are saying its not valid semver format but npm version command allows it, that is the tiny bit I am holding on to:)

The reason that I am doing this to be able to release from multiple branches. Here how it goes:

In each branch, including feature branches, I am running semantic-release with custom exec commands and publishing the artifact with packageName-semver-customPostfix  format and adding a dist tag with branch name to point my newly published component. By this way, I can always get the latest pakcage of my component from registry. During this process, I am not even bumping the version to package.json.

I just wanted to do this from CLI instead of needing to run exec commands to customize it. Did I make myself more clear and how is your opinion on this approach? Currently, it seems to work for us.

**Comment by pvdlg** [comment]
I'm not even sure how that works as `fdp-tool-cli-1.1.0-feature14` is not a valid semver, so I'm surprised npm even accepts it.

```js
const semver = require('semver');
semver.valid('fdp-tool-cli-1.1.0-feature14');
//=> null
```

In order to define the version published npm reads the `package.json` you send in the published package (the .tgz), but you mention you don't update it. So I'm not really what you are publishing exactly...If you don't update it you would publish the version `0.0.0-dev` over and over, which npm would refuse as a version can be published only once.

Maybe you can provide a link to your repo?

Anyway in any case, semantic-release will not implement a feature to publish invalid semver.

**Comment by tarazansafak** [comment]
I am using private registry on artifactory. I think npm accepts it because its pre-release format isn't it?:

`semver.valid(semver.coerce('42.6.7.9.3-alpha')) // '42.6.7'`.

Here is an example from my private registry and all the version info is kept in all.json on artifactory automagically.
fdp-tool-cli-1.1.3-CICD-282-semver-pilot.tgz

Anyways, you can close this issue. I will be also watching [563](https://github.com/semantic-release/semantic-release/issues/563). Thanks for your time.

**Comment by pvdlg** [comment]
`42.6.7.9.3-alpha` is a valid pre-release version indeed, but for info `fdp-tool-cli-1.1.0-feature14` is not.

**Comment by tarazansafak** [comment]
yes you are right, I just copied the whole string including the package name. Anyhow, it would be nice if we can support this pre-release with semantic-version plugin. By passing the pre-release part via cli as we are doing it with tag-format command.

**Comment by pvdlg** [comment]
That's going to be handled within #563 via the `branches` option
